django/db/models/query.py

```diff
@@ -366,6 +409,17 @@ def iterator(self, chunk_size=2000):
         use_chunked_fetch = not connections[self.db].settings_dict.get('DISABLE_SERVER_SIDE_CURSORS')
         return self._iterator(use_chunked_fetch, chunk_size)
 
+    async def aiterator(self, chunk_size=2000):
+        """
+        An asynchronous iterator over the results from applying this QuerySet
+        to the database.
+        """
+        if chunk_size <= 0:
+            raise ValueError('Chunk size must be strictly positive.')
+        use_chunked_fetch = not connections[self.db].settings_dict.get('DISABLE_SERVER_SIDE_CURSORS')
+        async for item in self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size):
```

Comment:
```python
async with contextlib.aclosing(self._iterable_class(...)) as agen:
    async for item in agen:
        yield agen
```

https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/#cleanup-in-generators-and-async-generators

Reply:
Does that matter here, though? `BaseIterable` and company do not rely on `__del__` or any other aspect of operator closure - or is there another side-effect I am unaware of?

Reply:
> We'll want to do something with regards to the newly added support for iterator's prefetch_related here.

That looks like a _moderate_ task in itself (to implement) — `islice`, `prefetch_related_objects`, ... — it might be that adjusting the PR here match the new interface, but emitting a warning if prefetches are set would let us get this in, to work on async prefetches later. (Would be equivalent to the sync behaviour before edbf930287cb72e9afab1f7208c24b1146b0c4ec — of _either prefetch or iterator_.)

Reply:
Agreed, we can raise `NotSupportedError` for now.